text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,353 @@
+- `TraceId` of the `Span` to be created.
+- Name of the `Span` to be created.
+- Initial set of `Attributes` of the `Span` to be created.
+- Collection of links that will be associated with the `Span` to be created.
+
+#### Required Arguments for RuleBased
+
+- `SpanKind`
+- list of pairs (`Predicate`, `Sampler`)
+
+For making the sampling decision, if the `Span` kind matches the specified kind, the sampler goes through the list in the provided order and calls `SpanMatches` on `Predicate`s passing the same arguments as received by `ShouldSample`. If a call returns `true`, the corresponding `Sampler` will be called to make the final sampling decision. If the `SpanKind` does not match, or none of the calls to `SpanMatches` yield `true`, the final decision is `DROP`.
+
+The order of `Predicate`s is essential. If more than one `Predicate` matches a `Span`, only the Sampler associated with the first matching `Predicate` will be used.
+
+## Summary - Approach One
+
+### Example - sampling configuration 1
+
+Going back to our example of sampling requirements, we can now configure the head sampler to support this particular case, using an informal notation of samplers and their arguments.
+First, let's express the requirements for the ROOT spans as follows.
+
+```
+S1 = RuleBased(ROOT, {
+ (http.target == /healthcheck) => AlwaysOff,
+ (http.target == /checkout) => AlwaysOn,
+ true => TraceIdRatioBased(0.25)
+ })
+```
+
+Note: technically, `ROOT` is not a `SpanKind`, but is a special token matching all Spans with invalid parent context (i.e. the ROOT spans, regardless of their kind).
+
+In the next step, we can build the sampler to handle non-root spans as well:
+
+```
+S2 = ParentBased(S1)
+```
+
+The special case of calling service `/foo` can now be supported by:
+
+```
+S3 = AnyOf(S2, RuleBased(CLIENT, { (http.url == /foo) => AlwaysOn })
+```
+
+Finally, the last step is to put a limit on the stream of exported spans. One of the available rate limiting sampler that we can use is Jaeger [RateLimitingSampler](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/RateLimitingSampler.java):
+
+```
+S4 = Conjunction(S3, RateLimitingSampler(1000 * 60))
+```
+
+### Example - sampling configuration 2
+
+Many users are interested in [Consistent Probability Sampling](https://opentelemetry.io/docs/specs/otel/trace/tracestate-probability-sampling/#consistent-probability-sampler) (CP), as it gives them a chance to calculate span-based metrics even when sampling is active. The configuration presented above uses the traditional samplers, which do not offer this benefit.
+
+Here is how an equivalent configuration can be put together using [CP samplers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#samplers). In this example, the following implementations are used:
```

Comment:
> EachOf(
>   RuleBased(ROOT, {IsTypeA => ConsistentFixedThreshold(0.4), IsTypeB => ConsistentFixedThreshold(0.2)}), 
>   ConsistentRateLimiting(750 * 60)
> )
> ```
> 
> This would give 250/min for A and 400/min for B, which gives a total of 650/min. Hence, the result differs from that of configuration 1.
> 
The total of 650/min is a concern, I'm less worried about the individual distribution between type A and B. After all, Consistent samplers have very different behavior than leaky-bucket (balancing vs. proportional).

Reply:
> 
> To make configuration 2 equivalent, we must replace `EachOf` with `Conjunction`. 

Yes, `Conjunction` will work better here. Which raises another question: do we need `EachOf` at all (neither in Approach One nor Approach Two)?

Reply:
> Furthermore, we would need a weighted variant of the `ConsistentRateLimiting` sampler, which analyzes the recent distribution of the `th` values given by the first sampler of the conjunction and which adapts the `th` values in such a way that the desired sampling rate is met. Ideally, the `ConsistentWeightedRateLimiting` would increase the thresholds for A and B from 0.6 to 0.675 and from 0.8 to 0.8375, respectively. (Assuming thresholds are values from [0,1].) This would again give 325/min for both, like configuration 1.
> 
Looking at the `th` values only makes sense if the given sampler is used as the Second delegate in `Conjunction`.
So how about generalizing this idea by defining `Weighted` or `Proportional` sampler which analyzes the population of the `rv` values in the population to sample. We cannot use the sampling probability value if it is influenced by the span's `rv`, but it does not mean that it cannot be influenced by the `rv` values from the _previous_ spans. Keeping track of the distribution of the `rv` values at a given sampling stage allows us not only to define a proportional variant of `RateLimitingSampler`, but also something like `ProportionalFixedRate` sampler. It would always reduce the volume of spans according to the defined factor. Not much use in head sampling perhaps, but I believe it would be useful in tail sampling.

More specifically, such a sampler would track the average `rv` of the incoming spans. For an unsampled population this average is expected to be 0.5. If the population has been pre-sampled, the spans with smaller `rv` would have been eliminated, and the observed average would be higher. This would allow the samplers with _proportional_ behavior to adjust their sampling probability accordingly (just as you proposed in your example).